### PR TITLE
Allow custom MySQL & Postgres JDBC connection string params

### PR DESCRIPTION
**Pull request by camsaul**
Tweak the way MySQL driver creates JDBC connection strings to allow custom parameters to be included. This will allow @hugocar to work around the problem he describes in #3506.

Includes tests

### CODE REVIEW

File: test/metabase/driver/mysql_test.clj [hunk 1]
```diff
@@ -12,6 +15,30 @@
...
+
+;; make sure if extra parameters are included that the correct combos of ?/& are used
+(expect
+  {:classname   "com.mysql.jdbc.Driver"
+   :subprotocol "mysql"
+   :subname     "//localhost:3306/cool?tinyInt1isBit=false&zeroDateTimeBehavior=convertToNull&useUnicode=true&characterEncoding=UTF8&characterSetResults=UTF8&useSSL=false"
+   :delimiters  "`"}
+  (generic-sql/connection-details->spec (MySQLDriver.) {:host   "localhost"
+                                                        :port   "3306"
+                                                        :dbname "cool?tinyInt1isBit=false"}))
```
Comment (salsakran) [inline test/metabase/driver/mysql_test.clj]: wait, are we going to have people enter `database names` with weird additional parameters tacked on to the end?

Comment (camsaul) [inline test/metabase/driver/mysql_test.clj]: One single guy who had a really bad schema is going to have to do it here to work around the default JDBC driver behavior.

**Comment by salsakran** [comment]
👎  on having people add connection parameters to the database name. 

If we have something that says "database name" on the label, it should just include the database name. I'm a 👍 on allowing connection string parameters to be passed in, but it should be a separate input.

**Comment by camsaul** [comment]
Alright, I'll close this PR and we can revisit at some point in the future 😞

**Comment by salsakran** [comment]
I think this is 80% there, and very useful. I'd keep it open and perhaps just add another input?

**Comment by camsaul** [comment]
I'll just switch this to add an extra field for "additional connection string params" or something like that if that works

**Comment by camsaul** [comment]
@salsakran @tlrobinson have another look. 

## Additional field

screen shot 2016-11-15 at 6 47 01 pm

## Default behavior:

screen shot 2016-11-15 at 6 56 12 pm

## With `tinyInt1isBit=false`

screen shot 2016-11-15 at 7 09 49 pm


Includes new tests.

**Comment by salsakran** [comment]
works for me. can this be generalized for other db types? I know we've had requests for passing mongo + postgresql connection string params as well.

**Comment by camsaul** [comment]
I added the additional connection args option for Postgres as well which means this PR also resolves issue #1899.

I looked into adding the options for Mongo as well, but we don't actually use a connection string internally for it; we'll have to look into some other approach. The only open issue we have that I could find requesting additional Mongo connection options was #1333, and it looks like someone submitted a PR to address it: #3727. We should take a look at that PR.

**Comment by salsakran** [comment]
looks good overall, one slightly minor nit, but will this be present in the setup process as well?

**Comment by camsaul** [comment]
yup

**Comment by edenzik** [comment]
@camsaul any way to do this for the Vertica driver as well? We need arbitrary connection strings for (1) disabling autocommit (2) enabling load balancing. If it's a trivial amount of work it'd help us a lot.

**Comment by camsaul** [comment]
Yeah that should be doable. I’ll take a look

**Comment by edenzik** [comment]
Thanks! Much appreciated.

**Pull request merged**